### PR TITLE
Default disguise for each InfantryType

### LINKED ISSUE #372: [New Feature] DefaultDisguise for individual TechnoTypes
### Description:
This would give the ability for each individual disguising infantry to has its own default disguise, which would override side-specific ones.  This could be useful if lets say the player was originally Allies then lost their base but captured Soviet tech.  It'd be pretty obvious that the first player is training spies when they come out as GIs.

### Possible Implementation:
```ini
[SOVSPY]
DefaultDisguise=E2 ; Defaults to [General]→AlliedDisguise, SovietDisguise, ThirdDisguise, or Ares' [SOMESIDE]→DefaultDisguise
```

### PR DESCRIPTION
**Pull request by secsome**
Closes #372 

### Default disguise for each InfantryType
- Infantry with `DefaultDisguise` can now has its own default disguise.
  - This tag's priority is higher than Ares' [Sides]>DefaultDisguise

In `rulesmd.ini`:
```ini
[SOMEINFANTRY]                    ; InfantryType
DefaultDisguise=E2                ; InfantryType              
```

### CODE REVIEW

**Comment by Thrifinesma** [comment]
What if this expands to per side like Ares' `Survivor.Side#=` or per country?

**Comment by secsome** [comment]
> What if this expands to per side like Ares' `Survivor.Side#=` or per country?

I think it should be implemented in another PR, leave it later

File: docs/New-or-Enhanced-Logics.md [hunk 1]
```diff
@@ -208,6 +208,16 @@ In `rulesmd.ini`:
...
 ```
 
+### Default disguise for each InfantryType
+- Infantry with `DefaultDisguise` can now has its own default disguise.
+  - This tag's priority is higher than Ares' [Sides]>DefaultDisguise
+
+In `rulesmd.ini`:
+```ini
+[SOMEINFANTRY]                    ; InfantryType
+DefaultDisguise=E2                ; InfantryType
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:218]:
```suggestion
[SOMEINFANTRY]      ; InfantryType
DefaultDisguise=E2  ; InfantryType              
```

File: docs/New-or-Enhanced-Logics.md [hunk 2]
```diff
@@ -208,6 +208,16 @@ In `rulesmd.ini`:
 NotHuman.RandomDeathSequence=yes  ; boolean
 ```
 
+### Default disguise for each InfantryType
+- Infantry with `DefaultDisguise` can now has its own default disguise.
+  - This tag's priority is higher than Ares' [Sides]>DefaultDisguise
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:213]:
```suggestion
### Default disguise for individual InfantryTypes

- Infantry can now have its `DefaultDisguise` overridden per-type.
  - This tag's priority is higher than Ares' per-side `DefaultDisguise`.
```

File: src/Ext/TechnoType/Hooks.cpp [hunk 3]
```diff
@@ -156,5 +157,25 @@ DEFINE_HOOK(0x700C58, TechnoClass_CanPlayerMove_NoManualMove, 0x6)
...
+{
+	GET(TechnoClass*, pThis, ESI);
+
+	if (auto const pExt = TechnoTypeExt::ExtMap.Find(pThis->GetTechnoType()))
+	{
+		if (pExt->DefaultDisguise.isset())
+		{
+			pThis->Disguise = pExt->DefaultDisguise;
+			pThis->Disguised = true;
+			return R->Origin() == 0x522790 ? 0x5227BF : 0x6F4277;
```
Comment (Metadorius) [inline src/Ext/TechnoType/Hooks.cpp:174]: The raw addresses that are used here are non-descriptive. Use anonymous enum to alias them.

**Review by Metadorius — CHANGES_REQUESTED** [review]
Code quality is good except the addresses part. Fix it - and then it will be mergable.

**Review by Metadorius — APPROVED** [review]

**Comment by DmitryVolkov666** [comment]
[SOMEINFANTRY] 
DefaultDisguise=

image1
image4
image3
image2

I (Volkov_T-666), as the owner, am red (My color)
YuriAL sees it as yellow (Own color)
Zazo4ek sees it as purple (Own color)

Although they both (YuriAL, Zazo4ek) should see it with red collor (My)

Moreover, if you try to disguise yourself in multiplayer as a foot soldier who already has a default disguise that doesn't work correctly, the game will crash referring to gamemd.exe and address (not immediately, but after a few tries).

But if we use it from ares:

[SOMESIDE]
DefaultDisguise=

There are no such problems. the owner of the object will be displayed correctly and there are no crashes when trying to disguise

image5

I hope I explained everything clearly. My translation is very bad

**Pull request merged**